myelin/kernel/cuda-array.cc

```diff
@@ -0,0 +1,492 @@
+    ptx_decl(pred, more);
+    ptx_emit(setp.lt.u32, more, fidx, PTXImm(num_features));
+    ptx_if(more);
+    ptx_jump(loop1);
+    ptx_endif();
+
+    // Save sum to output.
+    ptx_decl(b64, vptr);
+    ptx->LoadTensorAddress(vptr, v);
+    ptx_emit(mad.wide.u32, vptr, idx, PTXImm(v->stride(1)), vptr);
+    ptx_emit(st.global.f32, PTXAddr(vptr), sum);
+
+    // Done.
+    ptx_label(done);
+    ptx_ret();
+  }
+
+  int64 Complexity(const Step *step) override {
+    return step->input(0)->elements() * step->output(0)->elements();
+  }
+};
+
+// CUDA-based feature collect operation for recurrent features mapped through
+// an embedding matrix.
+class CUDACollect : public CUDAKernel {
+ public:
+  string Name() override { return "CUDACollect"; }
+  string Operation() override { return "Collect"; }
+
+  bool Supports(Step *step) override {
+    // Requires CUDA support.
+    if (!CUDAKernel::Supports(step)) return false;
+
+    // Check inputs and outputs.
+    if (step->indegree() != 2 || step->outdegree() != 1) return false;
+
+    // Check types.
+    Tensor *f = step->input(0);
+    Tensor *M = step->input(1);
+    Tensor *R = step->output(0);
+    if (f->type() != DT_INT32) return false;
+    if (M->type() != DT_FLOAT || M->rank() != 2) return false;
+    if (R->type() != DT_FLOAT || R->rank() != 2) return false;
+
+    if (f->dim(0) != 1 || f->dim(1) != R->dim(0)) return false;
+    if (R->dim(1) != M->dim(1) + 1) return false;
+
+    return true;
+  }
+
+  void Adjust(Step *step) override {
+    step->input(1)->SetRequiredOrder(ROW_MAJOR);
+    step->output(0)->SetRequiredOrder(ROW_MAJOR);
+  }
+
+  void GeneratePTX(Step *step, PTXMacroAssembler *ptx) override {
+    // Get inputs and outputs.
+    Tensor *f = step->input(0);
+    Tensor *M = step->input(1);
+    Tensor *R = step->output(0);
+
+    // Get size of activation vectors.
+    int dims = M->dim(1);
+
+    // Get number input features.
+    int num_features = f->dim(1);
+
+    // Use a thread for each feature and embedding dim.
+    ptx->set_grid_dims(dims, num_features);
+
+    // Get feature and embedding position thread index.
+    ptx_decl(b32, midx);
+    ptx->GetThreadIndex(midx, 0);
+    ptx_decl(b32, fidx);
+    ptx->GetThreadIndex(fidx, 1);
+
+    // Check bounds.
+    ptx_decl(pred, outside_x);
+    ptx_emit(setp.ge.u32, outside_x, midx, PTXImm(dims));
+    ptx_decl(pred, outside_y);
+    ptx_emit(setp.ge.u32, outside_y, fidx, PTXImm(num_features));
+    ptx_decl(pred, outside);
+    ptx_emit(or.pred, outside, outside_x, outside_y);
+    ptx_if(outside);
+    ptx_jump(done);
+    ptx_endif();
+
+    // Get feature id.
+    ptx_decl(b64, fptr);
+    ptx->LoadTensorAddress(fptr, f);
+    ptx_decl(s32, fval);
+    ptx_emit(mad.wide.u32, fptr, fidx, PTXImm(f->stride(1)), fptr);
+    ptx_emit(ld.global.s32, fval, PTXAddr(fptr));
+
+    // Output zero if feature value is -1.
+    ptx_decl(f32, value);
+    ptx_emit(mov.f32, value, PTXFloat(0));
```

Comment:
While that is technically true, it will still execute the same number of instructions, so it will not be faster. The instructions between ptx_if and ptx_endif will be conditionally executed based on the predicate value, but I think they will still take up an instruction cycle even when the predicate is false.